File: horton/moments.cpp
// HORTON: Helpful Open-source Research TOol for N-fermion systems.
// Copyright (C) 2011-2019 The HORTON Development Team
//
// This file is part of HORTON.
//
// HORTON is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version 3
// of the License, or (at your option) any later version.
//
// HORTON is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, see <http://www.gnu.org/licenses/>
//
//--


#include <stdexcept>
#include <cmath>
#include "horton/moments.h"


long fill_cartesian_polynomials(double* output, long lmax) {
    // Shell l=0
    if (lmax <= 0) return -1;

    // Shell l=1
    if (lmax <= 1) return 0;

    // Shell l>1
    int old_offset = 0; // first array index of the moments of the previous shell
    int old_ncart = 3;  // number of moments in previous shell
    for (int l = 2; l <= lmax; l++) {
        // numbers for current iteration
        int new_ncart = old_ncart + l+1;
        int new_offset = old_offset + old_ncart;
        // start by multiplying all the old ones by x
        for (int i=0; i < old_ncart; i++) {
            output[new_offset+i] = output[0]*output[old_offset+i];
        }
        // then multiply the old ones without an x, by y.
        for (int i=0; i < l; i++) {
            output[new_offset+old_ncart+i] = output[1]*output[new_offset-l+i];
        }
        // finally multiply the last old one by z
        output[new_offset+new_ncart-1] = output[2]*output[new_offset-1];
        // translate new to old numbers
        old_ncart = new_ncart;
        old_offset = new_offset;
    }
    return old_offset;
}


long fill_pure_polynomials(double* output, long lmax) {
    // Shell l=0
    if (lmax <= 0) return -1;

    // Shell l=1
    if (lmax <= 1) return 0;

    // Shell l>1

    // auxiliary variables
    double z = output[0];
    double x = output[1];
    double y = output[2];
    double r2 = x*x + y*y + z*z;

    // work arrays in which the PI(z,r) polynomials are stored.
    double pi_old[lmax+1];
    double pi_new[lmax+1];
    double a[lmax+1];
    double b[lmax+1];

    // initialize work arrays
    pi_old[0] = 1;
    pi_new[0] = z;
    pi_new[1] = 1;
    a[1] = x;
    b[1] = y;

    int old_offset = 0; // first array index of the moments of the previous shell
    int old_npure = 3;  // number of moments in previous shell
    for (int l = 2; l <= lmax; l++) {
        // numbers for current iteration
        int new_npure = old_npure + 2;
        int new_offset = old_offset + old_npure;

        // construct polynomials PI(z,r) for current l
        double factor = (2*l-1);
        for (int m=0; m<=l-2; m++) {
            double tmp = pi_old[m];
            pi_old[m] = pi_new[m];
            pi_new[m] = (z*factor*pi_old[m] - r2*(l+m-1)*tmp)/(l-m);
        }
        pi_old[l-1] = pi_new[l-1];
        pi_new[l] = factor*pi_old[l-1];
        pi_new[l-1] = z*pi_new[l];

        // construct new polynomials A(x,y) and B(x,y)
        a[l] = x*a[l-1] - y*b[l-1];
        b[l] = x*b[l-1] + y*a[l-1];

        // construct solid harmonics
        output[new_offset] = pi_new[0];
        factor = M_SQRT2;
        for (int m=1; m<=l; m++) {
            factor /= sqrt((l+m)*(l-m+1));
            output[new_offset+2*m-1] = factor*a[m]*pi_new[m];
            output[new_offset+2*m] = factor*b[m]*pi_new[m];
        }

        // translate new to old numbers
        old_npure = new_npure;
        old_offset = new_offset;
    }
    return old_offset;
}


long fill_pure_polynomials_array(double* output, long lmax, long nrep, long stride) {
    long result = 0;
    for (long irep=0; irep<nrep; irep++) {
        result = fill_pure_polynomials(output, lmax);
        output += stride;
    }
    return result;
}


void fill_radial_polynomials(double* output, long lmax) {
    // Shell l=0
    if (lmax <= 0) return;

    // Shell l=1
    if (lmax <= 1) return;

    // Shell l>1
    int counter = 0;
    while (lmax > 1) {
        output[counter+1] = output[0]*output[counter];
        lmax -= 1;
        counter += 1;
    }
}

File: tools/moments.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# HORTON: Helpful Open-source Research TOol for N-fermion systems.
# Copyright (C) 2011-2016 The HORTON Development Team
#
# This file is part of HORTON.
#
# HORTON is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 3
# of the License, or (at your option) any later version.
#
# HORTON is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, see <http://www.gnu.org/licenses/>
#
# --
"""Pre-computation and code generation for all sorts of moments"""


from horton import get_cartesian_powers, get_ncart


def write_cart_poly_code(f):
    cps = get_cartesian_powers(7)
    cps = [tuple(row) for row in cps[1:]] # chop of the s function

    print >>f, "    // Shell l=0"
    print >>f, "    if (lmax <= 0) return -1;"
    print >>f


    last_shell = 0
    begin = -1
    end = 0
    for px, py, pz in cps:
        shell = px+py+pz
        if shell > last_shell:
            print >>f, "    // Shell l=%i" % shell
            begin = end
            last_shell = shell

        if shell > 1:
            if shell > 2:
                if px != 0:
                    prev_row = (px-1, py, pz)
                    symbol = 'output[0]'
                elif py != 0:
                    prev_row = (px, py-1, pz)
                    symbol = 'output[1]'
                else:
                    prev_row = (px, py, pz-1)
                    symbol = 'output[2]'
                formula = '%s*output[%i]' % (symbol, cps.index(prev_row))
            else:
                symbols = ["output[0]"]*px + ["output[1]"]*py + ["output[2]"]*pz
                formula = "*".join(symbols)
            print >>f, "    output[%i] = %s;" % (end, formula)
        end += 1

        if end >= begin+get_ncart(shell):
            print >>f, "    if (lmax <= %i) return %i;" % (shell, begin)
            print >>f

    print >> f, '    throw std::domain_error("Encountered lmax > 7.");'


def write_cart_rotation_code(f):
    import numpy as np
    from sympy import Symbol, Wild, Add, Mul, Integer


    R = [Symbol('R_%d' % i) for i in xrange(9)]
    x = Symbol('x')
    y = Symbol('y')
    z = Symbol('z')

    lmax = 4
    cartesian_powers = [tuple(row) for row in get_cartesian_powers(lmax)]
    print cartesian_powers
    ncart = len(cartesian_powers)


    transforms = []

    def get_power(term, base):
        p = Wild("p")
        c = Wild("c")
        d = term.match(c*base**p)
        if d is None:
            return 0
        else:
            return int(d[p])

    icart = 0
    for ishell in xrange(lmax+1):
        shell_rules = []
        transforms.append(shell_rules)
        for ifn in xrange(((ishell+1)*(ishell+2))/2):
            px0, py0, pz0 = cartesian_powers[icart]
            rules = []
            shell_rules.append(rules)
            poly = (
                 (R[0]*x + R[1]*y + R[2]*z)**px0
                *(R[3]*x + R[4]*y + R[5]*z)**py0
                *(R[6]*x + R[7]*y + R[8]*z)**pz0
            ).expand()
            print poly

            if isinstance(poly, Add):
                terms = poly.args
            else:
                terms = [poly]

            for term in terms:
                px1 = get_power(term, x)
                py1 = get_power(term, y)
                pz1 = get_power(term, z)
                prs = [get_power(term, R[i]) for i in xrange(9)]
                col = cartesian_powers.index((px1, py1, pz1)) - cartesian_powers.index((px1+py1+pz1, 0, 0))
                assert col >= 0
                coeff = 1
                if isinstance(term, Mul):
                    for factor in term.args:
                        if isinstance(factor, Integer):
                            coeff *= int(factor)
                rule = (col, coeff, prs)
                rules.append(rule)

            icart += 1


    def format_rule(rule):
        l = [rule[0], rule[1]]
        prs = rule[2]
        for j in xrange(9):
            for p in xrange(prs[j]):
                l.append(j)
            #if prs[j] > 0:
            #    l.append(j)
            #    l.append(prs[j])
        return '[%s]' % (', '.join('%2i' % i for i in l))


    print >> f, 'cartesian_transforms = ['
    for shell_rules in transforms:
        print >> f, '  ['
        for rules in shell_rules:
            if len(rules) == 1:
                print >> f, '    [%s],' % format_rule(rules[0])
            else:
                rules = sorted(rules)
                print >> f, '    [%s,' % format_rule(rules[0])
                for rule in rules[1:-1]:
                    print >> f, '     %s,' % format_rule(rule)
                print >> f, '     %s],' % format_rule(rules[-1])
        print >> f, '  ],'
    print >> f, ']'


if __name__ == "__main__":
    #with file("cart_poly.inc.cpp", "w") as f:
    #    write_cart_poly_code(f)
    with file("cart_rotate.inc.py", "w") as f:
        write_cart_rotation_code(f)
